__tests__/authkit-provider.spec.tsx

```diff
@@ -0,0 +1,149 @@
+    expect(window.addEventListener).not.toHaveBeenCalled();
+  });
+
+  it('should call onSessionExpired when session is expired', async () => {
+    (checkSessionAction as jest.Mock).mockRejectedValueOnce(new Error('Failed to fetch'));
+    const onSessionExpired = jest.fn();
+
+    render(
+      <AuthKitProvider onSessionExpired={onSessionExpired}>
+        <div>Test Child</div>
+      </AuthKitProvider>,
+    );
+
+    // Simulate visibility change
+    window.dispatchEvent(new Event('visibilitychange'));
+
+    await waitFor(() => {
+      expect(onSessionExpired).toHaveBeenCalled();
+    });
+  });
+
+  it('should only call onSessionExpired once if multiple visibility changes occur', async () => {
+    (checkSessionAction as jest.Mock).mockRejectedValueOnce(new Error('Failed to fetch'));
+    const onSessionExpired = jest.fn();
+
+    render(
+      <AuthKitProvider onSessionExpired={onSessionExpired}>
+        <div>Test Child</div>
+      </AuthKitProvider>,
+    );
+
+    // Simulate visibility change twice
+    window.dispatchEvent(new Event('visibilitychange'));
+    window.dispatchEvent(new Event('visibilitychange'));
+
+    await waitFor(() => {
+      expect(onSessionExpired).toHaveBeenCalledTimes(1);
+    });
+  });
+
+  it('should pass through if checkSessionAction does not throw "Failed to fetch"', async () => {
+    (checkSessionAction as jest.Mock).mockResolvedValueOnce(false);
+
+    const onSessionExpired = jest.fn();
+
+    render(
+      <AuthKitProvider onSessionExpired={onSessionExpired}>
+        <div>Test Child</div>
+      </AuthKitProvider>,
+    );
+
+    // Simulate visibility change
+    window.dispatchEvent(new Event('visibilitychange'));
+
+    await waitFor(() => {
+      expect(onSessionExpired).not.toHaveBeenCalled();
+    });
+  });
+
+  it('should reload the page when session is expired and no onSessionExpired handler is provided', async () => {
+    (checkSessionAction as jest.Mock).mockRejectedValueOnce(new Error('Failed to fetch'));
+
+    const originalLocation = window.location;
+
+    // @ts-expect-error - we're deleting the property to test the mock
+    delete window.location;
+
+    window.location = { ...window.location, reload: jest.fn() };
+
+    render(
+      <AuthKitProvider>
+        <div>Test Child</div>
+      </AuthKitProvider>,
+    );
+
+    // Simulate visibility change
+    window.dispatchEvent(new Event('visibilitychange'));
+
+    await waitFor(() => {
+      expect(window.location.reload).toHaveBeenCalled();
+    });
+
+    // Restore original reload function
+    window.location = originalLocation;
+  });
+
+  it('should not call onSessionExpired or reload the page if session is valid', async () => {
+    (checkSessionAction as jest.Mock).mockResolvedValueOnce(true);
+    const onSessionExpired = jest.fn();
+
+    const originalLocation = window.location;
+
+    // @ts-expect-error - we're deleting the property to test the mock
+    delete window.location;
```

Comment:
could we maybe do something like `jest.spyOn(window.location, 'reload')` instead of replacing munging the window object directly? (then your location will be properly restored even if an exception is thrown/etc).

Reply:
Unfortunately not because `window.location.reload` is a read only property. Spying on it with jest still attempts to modify the function giving you a `TypeError: Cannot assign to read only property 'reload' of object '[object Location]'` error.